### ISSUE TITLE
Request for Webkit's position: Explicit compile hints

### ISSUE THREAD

**Issue by marjakh** [WebKit/standards-positions#172]
### Title of the spec

Explicit compile hints

### URL to the spec

https://docs.google.com/document/d/19xTAM4A75tz0xUq_velMzGA4JHEgXpyflUxXTcuNiyE/edit?usp=sharing

### Mozilla standards-positions issue URL

https://github.com/mozilla/standards-positions/issues/780

### Description

Knowing which JavaScript functions to parse & compile upfront can speed up web page loading.

Some example numbers from our experiments:
- Data set of 4 big web pages, local cold runs: Parse + compile time decreases 4 - 13 %, foreground parse + compile time decreases 59 - 86%.
- Internal web devs report 5% improvement in their page load time metrics w/ PIFEing all functions in core file.

Currently, V8 uses the [PIFE heuristic](https://v8.dev/blog/preparser#pife) to direct which functions to compile. This heuristic has existed for a long time and is well known to web developers. V8 and SpiderMonkey follow the PIFE heuristic, but JavaScriptCore doesn't.

Using PIFEs for transmitting information about which functions should be eager-compiled has downsides, though. Especially:
- using it forces using function expressions instead of function declarations; this might regress in all browsers, and for browsers which don't follow the PIFE hint there's no upside
- it cannot be applied to ES6 class methods

Thus, we want to experiment with other ways to transmit compile hints.

We don't want to cause extra work for browsers who don't want to support compile hints; we'll only explore options where ignoring the hints happens naturally.

We're currently starting early experiments with potential users, and will create the actual spec based on feedback.

**Comment by hober** [comment]
Added `concerns: venue` because, at first blush, it seems like this should be proposed in the context of TC39.

**Comment by marjakh** [comment]
We don't expect TC39 to be the right venue, since we don't plan to suggest semantic changes to the JavaScript language. Adding compile hints as magic comments or directives won't produce any observable differences (aside performance).

E.g., magic comments for source mapping URLs ( https://sourcemaps.info/spec.html )
```
//# sourceMappingURL=<url>
```
 are also not specced by TC39.

**Comment by msaboff** [comment]
I agree with @hober.  Why do we think that TC39 is NOT the right venue?

If this is a benefit browsers, why don't we think it would benefit other JavaScript code bases like Node?

At a minimum, a proposal should be given to TC39 to see if there is broader interest.

**Comment by syg** [comment]
> Why do we think that TC39 is NOT the right venue?

Primarily because, as this has no observable semantics for JS, I'm not sure what there is to seek consensus for. We can certainly present it as a call for collaboration and to gauge interest, however.

I hear you on the IPR and governance concerns, perhaps WICG?

**Comment by RByers** [comment]
Note there's an [Intent to Ship for this in Chromium](https://groups.google.com/a/chromium.org/g/blink-dev/c/uNUWZ14SYJI/m/7-_h1rCXAQAJ). Definitely we prefer to see this in a venue with good IPR etc, but we need a signal of interest from another party in order to do that. Is WebKit interested?

**Comment by kmiller68** [comment]
I don't have enough opinions yet to say we'd want to / would consider shipping this too. I mainly worry that anything easily generated by hand and not a PGO tool is too easy to misuse. I also worry that either way the set of things picked will be tuned for the dev's favorite browser and be suboptimal for other browsers.

That said, I have a few drive-by thoughts from looking at the proposed spec:

* It seems like some kind of place for a version would be good practice. 
* IMO, https://explainers-by-googlers.github.io/explicit-javascript-compile-hints-file-based/#extract-compile-hint-annotation should bail if any non-comment (I guess including `#!`) are detected before the hint. Putting a hint partway down the file and expecting it to work would be weird.
* Why allow both `@` and `#` to prefix the comment https://explainers-by-googlers.github.io/explicit-javascript-compile-hints-file-based/#matching-compile-hint-annotation? Seems like just picking one would be better. (The explainer uses `#` so I guess that seems like logical/existing choice)

**Comment by syg** [comment]
The current plan is WICG. It's a poor fit as a TC39 **proposal** since it has no observable semantics. It will be presented at TC39 this week though as an info session.

**Comment by marjakh** [comment]
FYI Chromium is now shipping the per-file version of this proposal and we're working on the per-function one. The explainer at https://github.com/WICG/explicit-javascript-compile-hints-file-based/ has been updated to reflect this.

One of the criticism towards the per-file compile hints was that it's too coarse grained and easy to use excessively. The per-function compile hints are finer grained and also require selecting individual functions, which the users will hopefully do thoughtfully via PGO.
